internal/resource/buffer.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resource
+
+import (
+	"errors"
+	"io"
+	"sync"
+)
+
+var (
+	ErrOutOfMemory    = errors.New("unable to grow buffer")
+	ErrOffsetTooLarge = errors.New("cannot seek past end of buffer")
+	ErrUnknownWhence  = errors.New("unknown whence")
+)
+
+type buffer struct {
+	buf         []byte
+	cursor      int64
+	end         int64
+	cap         int64
+	writeLock   sync.Mutex
+	writeAtLock sync.Mutex
+}
+
+func NewBuffer() *buffer {
```

Comment:
Usually these take an initial byte slice. Also, instead of returning a non-exported type you might have an interface representing your custom creation here (interface resource.Buffer of Reader, Seeker, Writer, WriteAt). `NewBuffer(buf []byte) Buffer`